covid19dp_submission/steps/download_analysis.py

```diff
@@ -0,0 +1,127 @@
+# Copyright 2022 EMBL - European Bioinformatics Institute
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import urllib
+
+import requests
+from ebi_eva_common_pyutils.logger import logging_config
+
+logger = logging_config.get_logger(__name__)
+
+
+def download_analysis(project, num_analysis, processed_analysis_file, download_target_dir):
+    total_analysis = total_analysis_in_project(project)
+    logger.info(f"total analysis in project {project}: {total_analysis}")
+
+    analysis_array = get_analysis_to_process(project, num_analysis, total_analysis, processed_analysis_file)
+    logger.info(f"number of analysis to process: {len(analysis_array)}")
+
+    download_files(analysis_array, download_target_dir, processed_analysis_file)
+
+
+def total_analysis_in_project(project):
+    count_url = f"https://www.ebi.ac.uk/ena/portal/api/filereportcount?accession={project}&result=analysis"
+    response = requests.get(count_url)
+    if response.status_code != 200:
+        logger.error(f"Error fetching total analysis count for project {project}")
+        response.raise_for_status()
+    return response.json()
+
+
+def get_analysis_to_process(project, num_analysis, total_analysis, processed_analysis_file):
+    offset = 0
+    limit = num_analysis
+    processed_analysis = get_processed_analysis(processed_analysis_file)
+    analysis_for_processing = []
+    while offset < total_analysis:
+        analysis_from_ena = get_analysis_from_ena(project, offset, limit)
+        unprocessed_analysis = filter_out_processed_analysis(analysis_from_ena, processed_analysis)
+
+        if len(analysis_for_processing) + len(unprocessed_analysis) >= num_analysis:
+            analysis_for_processing = analysis_for_processing + unprocessed_analysis[
+                                                                :(num_analysis - len(analysis_for_processing))]
+            break
+        else:
+            analysis_for_processing = analysis_for_processing + unprocessed_analysis
+            offset = offset + limit
+
+    return analysis_for_processing
+
+
+def get_analysis_from_ena(project, offset, limit):
```

Comment:
Suggest annotating with retry and jitter as suggested in the stand-up meeting today...

Reply:
added